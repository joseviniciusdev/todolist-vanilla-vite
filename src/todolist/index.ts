const setTemplate = () => {
	document.querySelector<HTMLDivElement>("#app")!.innerHTML = `
	<div class="tudo">
		<form class="area-input" data-submit>
			<input required data-input placeholder="Nova tarefa">
			<button>Adicionar</button>
		</form>
		<div class="list" data-tudo-tasks>
		</div>
	</div>
`;
}

export type state = {
	id: number,
	name: string,
	finish: boolean
}

export let state: state[] = []

let idInteraction = 0

const task = (name: string, id: number, finish: boolean) => `<div class="task ${finish ? 'finish' : ''}" >
<span class="name">${name}</span>
<div class="button" data-finish data-task="${id}" role="button">${finish ? 'Reativar' : 'Finalizar'}</div>
<div class="button" data-remove data-task="${id}" role="button">Remover</div>
</div>`

const remove = (id: number) => {
	state = state.filter(item => item.id !== id)

	return state

}

const changeStatus = (id: number) => {
	state = state.map(item => {
		if (item.id == id) {
			const a = item
			a.finish = !a.finish
			return a
		}

		return item
	})

	return state
}

const addEventRemove = () => {
	document.querySelectorAll('[data-remove]').forEach(item => {
		item.addEventListener('click', () => {
			remove(Number(item.getAttribute('data-task')))
			listTasks(state)
		})
	})
}

const addEventFinish = () => {
	document.querySelectorAll('[data-finish]').forEach(item => {
		item.addEventListener('click', () => {
			changeStatus(Number(item.getAttribute('data-task')))
			listTasks(state)
		})
	})
}

const listTasks = (state: state[]) => {

	const template = state.map(item => task(item.name, item.id, item.finish)).join('')
	document.querySelector('[data-tudo-tasks]')!.innerHTML = template
	addEventRemove()
	addEventFinish()
}

const appendTask = (state: state[], name: string):void =>  {
	idInteraction += 1

	state.push({
		id: idInteraction,
		name,
		finish: false
	})

	listTasks(state)
}

const submitForm = () => {
	document.querySelector<HTMLFormElement>('[data-submit]')!.addEventListener('submit', (e) => {
		e.preventDefault()
		const inputTask = document.querySelector<HTMLInputElement>('[data-input]')!

		const { value: name } = inputTask
		inputTask.value = ''

		if (name.trim()) {
			appendTask(state, name)
		}
	})
}

export {
	task,
	setTemplate,
	submitForm,
	remove,
	changeStatus,
	// events
	listTasks
}
