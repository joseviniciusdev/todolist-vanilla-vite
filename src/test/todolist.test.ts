import { describe, it, expect, beforeEach } from "vitest";

import { state, task, remove, changeStatus, setTemplate, submitForm } from '../todolist'
import { JSDOM } from "jsdom";
import { mockState } from "./mock/state";

describe('Tudo list', () => {
	it('Deve remover a tarefa do carrinho retornando um Array vazio', () => {
		let state: state[] = mockState
		remove(10)
		expect(state).not.toHaveLength
	})

	it('Deve alterar o status de "finish" da tarefa para "true"', () => {
		let state: state[] = mockState
		changeStatus(10)
		expect(state[0].finish).toEqual(false)
	})

	it('Deve retornar nova template', () => {
		let state: state[] = mockState

		const templateTasks = state.map(item => task(item.name, item.id, item.finish)).join('')

		expect(templateTasks).toMatchSnapshot()
	})
})

describe('Tudo list events DOM', () => {
	beforeEach(() => {
		const dom = new JSDOM()
		global.document = dom.window.document
		document.body.innerHTML = '<div id="app"></div>'
	})

	it('Deve adicionar a template inicial', () => {
		expect(document.body).toMatchSnapshot()
	})

	it('Deve fazer o fluxo de tarefas', () => {
		setTemplate()
		submitForm()

		document.querySelector('input')!.value = 'New Task'
		document.querySelector('button')!.click()
		expect(document.body).toMatchSnapshot()
		// expect()
		document.querySelector<HTMLButtonElement>('[data-finish]')!.click()
		expect(document.body).toMatchSnapshot()
		document.querySelector<HTMLButtonElement>('[data-remove]')!.click()
		expect(document.body).toMatchSnapshot()
	})
})
