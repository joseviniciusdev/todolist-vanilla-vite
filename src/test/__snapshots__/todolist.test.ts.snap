// Vitest Snapshot v1

exports[`Tudo list > Deve retornar nova template 1`] = `
"<div class=\\"task \\" >
<span class=\\"name\\">exemplo</span>
<div class=\\"button\\" data-finish data-task=\\"10\\" role=\\"button\\">Finalizar</div>
<div class=\\"button\\" data-remove data-task=\\"10\\" role=\\"button\\">Remover</div>
</div><div class=\\"task finish\\" >
<span class=\\"name\\">exemplo 2</span>
<div class=\\"button\\" data-finish data-task=\\"20\\" role=\\"button\\">Reativar</div>
<div class=\\"button\\" data-remove data-task=\\"20\\" role=\\"button\\">Remover</div>
</div>"
`;

exports[`Tudo list events DOM > Deve adicionar a template inicial 1`] = `
<body>
  <div
    id="app"
  />
</body>
`;

exports[`Tudo list events DOM > Deve fazer o fluxo de tarefas 1`] = `
<body>
  <div
    id="app"
  >
    
	
    <div
      class="tudo"
    >
      
		
      <form
        class="area-input"
        data-submit=""
      >
        
			
        <input
          data-input=""
          placeholder="Nova tarefa"
          required=""
        />
        
			
        <button>
          Adicionar
        </button>
        
		
      </form>
      
		
      <div
        class="list"
        data-tudo-tasks=""
      >
        <div
          class="task "
        >
          

          <span
            class="name"
          >
            New Task
          </span>
          

          <div
            class="button"
            data-finish=""
            data-task="0"
            role="button"
          >
            Finalizar
          </div>
          

          <div
            class="button"
            data-remove=""
            data-task="0"
            role="button"
          >
            Remover
          </div>
          

        </div>
      </div>
      
	
    </div>
    

  </div>
</body>
`;

exports[`Tudo list events DOM > Deve fazer o fluxo de tarefas 2`] = `
<body>
  <div
    id="app"
  >
    
	
    <div
      class="tudo"
    >
      
		
      <form
        class="area-input"
        data-submit=""
      >
        
			
        <input
          data-input=""
          placeholder="Nova tarefa"
          required=""
        />
        
			
        <button>
          Adicionar
        </button>
        
		
      </form>
      
		
      <div
        class="list"
        data-tudo-tasks=""
      >
        <div
          class="task finish"
        >
          

          <span
            class="name"
          >
            New Task
          </span>
          

          <div
            class="button"
            data-finish=""
            data-task="0"
            role="button"
          >
            Reativar
          </div>
          

          <div
            class="button"
            data-remove=""
            data-task="0"
            role="button"
          >
            Remover
          </div>
          

        </div>
      </div>
      
	
    </div>
    

  </div>
</body>
`;

exports[`Tudo list events DOM > Deve fazer o fluxo de tarefas 3`] = `
<body>
  <div
    id="app"
  >
    
	
    <div
      class="tudo"
    >
      
		
      <form
        class="area-input"
        data-submit=""
      >
        
			
        <input
          data-input=""
          placeholder="Nova tarefa"
          required=""
        />
        
			
        <button>
          Adicionar
        </button>
        
		
      </form>
      
		
      <div
        class="list"
        data-tudo-tasks=""
      />
      
	
    </div>
    

  </div>
</body>
`;
